Add a flow for returning data early from Mutation#authorized?

rmosolgo opened the pull request:
As pointed out by @cainlevy in #1734 , the new mutation auth doesn't support errors-as-data very well, but it definitely _should_, because errors-as-data is nice. 

How to do it? I can think of a few implementations: 

- Somehow use the return value of `.authorized?` (this is how before-save callbacks used to work, if they returned `false`, then save was prevented. I think this was changed to [`throw(:abort)`](https://blog.bigbinary.com/2016/02/13/rails-5-does-not-halt-callback-chain-when-false-is-returned.html)
- Use `throw :early_return, response`. This has the advantage of being plain Ruby, but it's not a very familiar feature. Also, it halts program flow, which is convenient, since later code will be irrelevant. 
- Use a normal method, as shown here. It's abstract, which is nice, but doesn't halt flow. This is like Rails controller `render` methods.
- Use a method that calls `throw` under the hood (like sinatra's `halt`, I think). It's nice that it halts flow, but maybe too magical? 

Any thoughts? 

TODO: 

- [ ] before_prepare should get the same treatment, be renamed to a `?` method.

cainlevy commented:
Thoughts on creating an official GraphQL::Result class? I imagine it's no small undertaking and would trickle out to many bits of the API, but it seems like the closest way to represent errors as data.

One option might be to always expect a Result back from `.authorized?`. The developer would need to indicate when it's a "continue" Result and when it's a "halt" result. Honestly this just feels like a Promise.

Another option would be to continue for any truthy value that is _not_ a `GraphQL::Result` object. In that case, `.authorized?` would need to return `GraphQL::Result.new(data)` when it wants to halt resolution with error data.

rmosolgo commented:
>  any truthy value that is _not_

Oh, that's a nice direction, something similar would be: 

- Continue on `true`
- Halt on:
  - `false`, with no data at all 
  - a non-`true` value, using that value as the response 

In that way, returning a hash would Just Work :tm:

cainlevy commented:
I'm inclined to caution in Ruby when there's a difference between `true` and truthy (or between `false` and falsey). I've had co-workers who thought it was an anti-pattern to use `!!` for casting a result to a strict boolean.

rmosolgo commented:
Oh yeah, I saw a rubocop against it once! 🤷‍♂️

rmosolgo commented:
I updated this to use multiple return for errors-as-data, eg 

```ruby
return false, { errors: [...] }
```

I also updated the guides to give a sense of it. How does that look to you?

cainlevy commented on lib/graphql/schema/resolver.rb:
```diff
@@ -210,6 +222,23 @@ def load_application_object_failed(err)
+      #
+      # @param response [Object] The value to respond with
+      # @return [void]
+      def return_with(response)
```
was this a previous idea? if it's still on the table, i think `resolve_with` might fit the GraphQL nomenclature nicely.

rmosolgo replied:
oops, oh yeah, it can be removed now since `return` is used instead
